Libraries/libbom/Sources/bom_tree.c

```diff
@@ -202,19 +213,28 @@ bom_tree_add(struct bom_tree_context *tree_context, const void *key, size_t key_
     int tree_index = bom_variable_get(tree_context->context, tree_context->variable_name);
     struct bom_tree *tree = (struct bom_tree *)bom_index_get(tree_context->context, tree_index, NULL);
 
-    /* Make room for the new index, extending the size as necessary. */
     int paths_index = ntohl(tree->child);
-    bom_index_append(tree_context->context, paths_index, sizeof(struct bom_tree_entry_indexes));
+    bool resize;
+    if (tree_context->preallocated_index_count > 0) {
```

Comment:
Nit:  It might be useful to abstract this logic out to something like "tryConsumePreallocatedResource" that encapsulates interacting with the index count.